src/timer.rs

```diff
@@ -0,0 +1,196 @@
+    }
+
+    pub fn cancel(&mut self, schedule_token: ScheduleToken) {
+        let _ = self.thread_tx.send(ThreadMessage::Cancel(schedule_token));
+    }
+
+    pub(crate) fn shutdown(&mut self) {
+        let _ = self.thread_tx.send(ThreadMessage::Shutdown);
+    }
+}
+
+impl TimerHandle {
+    pub fn new(timer_resolution: Duration) -> Self {
+        let (thread_tx, thread_rx) = crossbeam_channel::unbounded();
+
+        let mut timer_thread = TimerThread::new(timer_resolution, thread_rx);
+
+        let join_handle = std::thread::spawn(move || {
+            timer_thread.run();
+        });
+
+        let timer_ref = TimerRef { thread_tx };
+
+        Self { join_handle, timer_ref }
+    }
+
+    pub fn timer_ref(&self) -> TimerRef {
+        self.timer_ref.clone()
+    }
+
+    pub fn shutdown(self) {
+        let _ = self.timer_ref.thread_tx.send(ThreadMessage::Shutdown);
+        let _ = self.join_handle.join();
+    }
+}
+
+pub struct TimerThread {
+    timer_resolution: Duration,
+    thread_rx: Receiver<ThreadMessage>,
+    entries: Vec<TimerEntry>,
+}
+
+impl TimerThread {
+    fn new(timer_resolution: Duration, thread_rx: Receiver<ThreadMessage>) -> Self {
+        Self { timer_resolution, thread_rx, entries: Vec::new() }
+    }
+
+    fn run(&mut self) {
+        loop {
+            // Try to timer entries or control messages from the channel.
+            if let Ok(msg) = self.thread_rx.try_recv() {
+                match msg {
+                    ThreadMessage::Shutdown => {
+                        break;
+                    },
+                    ThreadMessage::Cancel(schedule_token) => {
+                        if let Some(pos) =
+                            self.entries.iter().position(|e| e.schedule_token == schedule_token)
+                        {
+                            self.entries.swap_remove(pos);
+                        }
+                    },
+                    ThreadMessage::Run(entry) => {
+                        self.entries.push(entry);
+                    },
+                }
+            }
+
+            // Expire and run all timers
+            let now = Instant::now();
+
+            for i in (0..self.entries.len()).rev() {
+                if now > self.entries[i].scheduled_at {
+                    let entry = self.entries.swap_remove(i);
+                    let schedule_token = entry.schedule_token;
+
+                    match entry.timer_type {
+                        TimerType::OneShot(closure) => {
+                            closure(schedule_token);
+                        },
+                        TimerType::Recurring(interval, mut closure) => {
+                            match closure(schedule_token) {
+                                TimerControlFlow::Continue => {
+                                    let new_entry = TimerEntry {
+                                        schedule_token,
+                                        scheduled_at: entry.scheduled_at + interval,
```

Comment:
Do we need to add code to check that the runtime wasn't so long that this `scheduled_at` would actually be in the past?

Reply:
If we cared a bit more about timer precision then yes, but in the case you describe, it should be picked up and run in the next loop iteration.

We could definitely make improvements to measure elapsed time after doing the task work and sleep less in that case, but that veers from the intended simplicity I had for this timer.